### PR TITLE
Instrumentation state

### PR DESCRIPTION
**Pull request by tmat**
In preparation for adding another instrumenter.

Move mutable instrumentation state maintained during lowering to a separate type.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/CompoundInstrumenter.cs [hunk 1]
```diff
@@ -24,6 +24,8 @@ public CompoundInstrumenter(Instrumenter previous)
 
         public Instrumenter Previous { get; }
 
+        public abstract CompoundInstrumenter WithPrevious(Instrumenter previous);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/CompoundInstrumenter.cs:27]: >WithPrevious [](http://example.com/codeflow?start=45&length=12)

Consider adding a doc comment explaining what this method is supposed to do

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/InstrumentationState.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+/// <summary>
+/// Manages instrumentation state.
+/// </summary>
+internal sealed class InstrumentationState
+{
+    /// <summary>
+    /// Used to temporary suspend instrumentation, for example when lowering expression tree.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/InstrumentationState.cs:13]: >temporary [](http://example.com/codeflow?start=16&length=9)

Perhaps "temporarily" or alternatively remove word.

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs [hunk 3]
```diff
@@ -22,13 +22,19 @@ internal partial class DebugInfoInjector : CompoundInstrumenter
...
+        private DebugInfoInjector(Instrumenter previous)
             : base(previous)
         {
         }
 
+        public static DebugInfoInjector Create(Instrumenter previous)
+            => (previous == NoOp) ? s_singleton : new DebugInfoInjector(previous);
+
+        public override CompoundInstrumenter WithPrevious(Instrumenter previous)
+            => Create(previous);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs:36]: >Create(previous) [](http://example.com/codeflow?start=15&length=16)

Should this method return `this` if `(object)previous == Previous`?

File: src/Compilers/CSharp/Portable/Lowering/Instrumentation/Instrumenter.cs [hunk 4]
```diff
@@ -5,6 +5,7 @@
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/Instrumentation/Instrumenter.cs:8]: >Roslyn.Utilities [](http://example.com/codeflow?start=6&length=16)

Is this change needed?

**Review by cston — APPROVED** [review]

**Comment by tmat** [comment]
@AlekseyTs Will follow up with VB change.

**Comment by tmat** [comment]
Actually, VB does not need the state changes. It passes the instrumenter around differently and doesn't need to turn off test coverage instrumentation for lambdas. So just adding the rename of `CodeCoverageInstrumenter` here.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs [hunk 5]
```diff
@@ -63,8 +63,11 @@ public override BoundNode VisitConversion(BoundConversion node)
 
             bool wasInExpressionLambda = _inExpressionLambda;
             _inExpressionLambda = _inExpressionLambda || (node.ConversionKind == ConversionKind.AnonymousFunction && !wasInExpressionLambda && rewrittenType.IsExpressionTree());
+            InstrumentationState.IsSuppressed = _inExpressionLambda;
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs:66]: >InstrumentationState.IsSuppressed = _inExpressionLambda; [](http://example.com/codeflow?start=12&length=56)

Should this change be tested?

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs (hunk 5 above)
Comment (tmat) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs:66]: This PR doesn't change the behavior. I'd hope there is a test covering this.
